spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/PayloadSerdeUtils.java

```diff
@@ -0,0 +1,230 @@
+															Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		Object originalPayload = messageValues.getPayload();
+		MimeType contentType = contentTypeResolver.resolve(new MessageHeaders(messageValues.getHeaders()));
+		Object payload = deserializePayload(originalPayload, contentType, payloadTypeCache, codec);
+		if (payload != null) {
+			messageValues.setPayload(payload);
+			Object originalContentType = messageValues.get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
+			// Reset content-type only if the original content type is not null (when
+			// receiving messages from
+			// non-SCSt applications).
+			if (originalContentType != null) {
+				messageValues.put(MessageHeaders.CONTENT_TYPE, originalContentType);
+				messageValues.remove(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
+			}
+		}
+		return messageValues;
+	}
+
+	private static Object deserializePayload(Object payload, MimeType contentType,
+											Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		if (payload instanceof byte[]) {
+			if (contentType == null || MimeTypeUtils.APPLICATION_OCTET_STREAM.equals(contentType)) {
+				return payload;
+			}
+			else {
+				return deserializePayload((byte[]) payload, contentType, payloadTypeCache, codec);
+			}
+		}
+		return payload;
+	}
+
+	private static Object deserializePayload(byte[] bytes, MimeType contentType,
+									Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		if ("text".equalsIgnoreCase(contentType.getType()) || MimeTypeUtils.APPLICATION_JSON.equals(contentType)) {
+			try {
+				return new String(bytes, "UTF-8");
+			}
+			catch (UnsupportedEncodingException e) {
+				String errorMessage = "unable to deserialize [java.lang.String]. Encoding not supported. "
+						+ e.getMessage();
+				PayloadSerdeUtils.LOGGER.error(errorMessage);
+				throw new SerializationFailedException(errorMessage, e);
+			}
+		}
+		else {
+			String className = JavaClassMimeTypeConversion.classNameFromMimeType(contentType);
+			try {
+				// Cache types to avoid unnecessary ClassUtils.forName calls.
+				Class<?> targetType = payloadTypeCache.get(className);
+				if (targetType == null) {
+					targetType = ClassUtils.forName(className, null);
+					payloadTypeCache.put(className, targetType);
+				}
+				return codec.decode(bytes, targetType);
+			} // catch all exceptions that could occur during de-serialization
+			catch (Exception e) {
+				String errorMessage = "Unable to deserialize [" + className + "] using the contentType [" + contentType
+						+ "] " + e.getMessage();
+				PayloadSerdeUtils.LOGGER.error(errorMessage);
+				throw new SerializationFailedException(errorMessage, e);
+			}
+		}
+	}
+
+	/**
+	 * Handles representing any java class as a {@link MimeType}.
+	 *
+	 * @author David Turanski
+	 * @author Ilayaperumal Gopinathan
+	 */
+	public abstract static class JavaClassMimeTypeConversion {
```

Comment:
If the class needs to be publicly accessible, then its static methods need to be accessible as well. I would move this to top level to avoid `PayloadSerdeUtils.JavaClassMimeTypeConversion.mimeTypeFromObject()` and I would change its name to something else - `JavaClassMimeTypeUtils` or so.